### PR TITLE
Average slice emittance - libEnsemble example

### PR DESCRIPTION
**Pull request by LDAmorim**
Hi all,

As Jean-Luc was mentioning during the Monday meeting, it might make more sense to use the beam average longitudinal slice normalized emittance (formula is listed below) instead of the projected one.

<p align="left">
    
</p>

Where <> represent the beam weighted averages of the quantities.

This PR implements this in the code.
The projected emittance can also be achieved by reducing the nr of slices in which the beam is divided to `nslices=1`.

Changes done:

*  `Tools/LibEnsemble/read_sim_output.py` - added ``slice_emittance`` function and used its average; fixed energy calculations to use particle weights and electron rest mass from `scipy`. I also changed averages to be weighed just so users know to take care of weights once they expand this example.
And added the safe alternative to when runs do not finish below the ``sim_kill_minutes`` defined in `run_libensemble_on_warpx.py`.
* ``Tools/LibEnsemble/warpx_simf.py`` - changed the `try` option because it didn't cover the case of no charge in the end of the simulation, i.e. beam is defocused (which can happen - specially when moving to 3D).
* ``Tools/LibEnsemble/plot_results.py`` - just to have energy spread in % units.

Cheers,
Diana

### CODE REVIEW

File: Tools/LibEnsemble/read_sim_output.py [hunk 1]
```diff
@@ -37,10 +45,22 @@ def _beam_properties(filepath):
     # Defined like that, the beam charge is > 0.
     charge = np.sum(w) * scc.e
     gamma = np.sqrt(1. + ux**2 + uy**2 + uz**2)
-    energy_MeV = .511 * (gamma - 1.)
-    energy_avg = np.mean(energy_MeV)
-    energy_std = np.std(energy_MeV) / energy_avg
-    emittance = np.sqrt(np.std(x)**2 * np.std(ux)**2 - np.mean(x*ux)**2)
+    beta = beta = np.sqrt(1.0 - 1.0 / gamma**2)
```
Comment (Yin-YinjianZhao) [inline Tools/LibEnsemble/read_sim_output.py]: Two `beta` ?

Comment (LDAmorim) [reply Tools/LibEnsemble/read_sim_output.py]: Thanks Yin! I will fix that :)

**Comment by LDAmorim** [comment]
The latests corrections were tested in this jupyter notebook.
[TestPREmittance.pdf](https://github.com/ECP-WarpX/WarpX/files/4904400/TestPREmittance.pdf)

**Comment by LDAmorim** [comment]
Before the latest changes my runs where crashing with an error because `warpx_out[0]` wasn't defined.
The recent changes allowed me to run the example in Summit @ OLCF and get the results below:

results

This plot shows the different simulations, each colored according to their ID # correlated to when they started running and with the same color for all plots.
Each column (and vertical axis) shows the output parameters that are relevant for this test case: f (parameter that we set and wish to minimize), emittance, charge, energy average and std.
Each row (and horizontal axis) shows the different input parameters that vary between all simulations: ramp_down_1, zlens_1, adjust_factor and ramp_down_2.


```
$ python plot_results.py /ccs/home/ligiada/scratch/LibEnsemble/Test/TestPR/
Reading latest .npy and .pickle files in /ccs/home/ligiada/scratch/LibEnsemble/Test/TestPR/
['/ccs/home/ligiada/scratch/LibEnsemble/Test/TestPR/run_libensemble_on_warpx_history_length=400_evals=400_workers=24.npy', '/ccs/home/ligiada/scratch/LibEnsemble/Test/TestPR/libE_history_for_run_starting_2020-07-13_16:29:12.903_after_sim_400.npy']
Best run:
ramp_down_1 [0.002]
zlens_1 [0.01274941]
adjust_factor [0.59158904]
ramp_down_2 [0.00531938]
f 2.9862198554191867e-07
emittance [2.98621986e-07]
charge [1.427145e-06]
energy_avg [104.05690638]
energy_std [0.03810686]
charge_f/charge_i  [1.]
```

That can be compared with the result with projected emittance:
```
ramp_down_1 [0.00221308]
zlens_1 [0.01244422]
adjust_factor [0.58478061]
ramp_down_2 [0.00498506]
f 4.884469342562086e-07
emittance [4.88446934e-07]
charge [1.427145e-06]
energy_avg [104.04666096]
energy_std [0.01913203]
charge_f/charge_i  [1.]
```

**Comment by ax3l** [comment]
Is it possible you get the NaNs from sqrt(<=0) values? Do you think it would be useful to filter such arguments?

**Comment by LDAmorim** [comment]
I get NaNs on purpose when some runs exceed the time limit and don't finish (i.e. there are no final plotfile) or when the beam is gone (i.e. no beam in final plotfile).
On purpose because that is a good way to predict that we don't want to include their results as they are far from optimal.
This is specially an issue for 3D runs, sensitive to the changes to the ramp sizes etc.

**Comment by LDAmorim** [comment]
@ax3l I think also that the advantage of using NANs (as Maxence was using also before) is due to the different files being called at different stages by libEnsemble and needing to store the results for each run in the final ones.
Then all we do at the end is filter out the NANs by using the `argmin/max` of `numpy` with `nan`.

File: Tools/LibEnsemble/warpx_simf.py [hunk 2]
```diff
@@ -85,12 +85,12 @@ def run_warpx(H, persis_info, sim_specs, libE_info):
...
-        warpx_out = read_sim_output(task.workdir)
-    except Exception:
-        warpx_out = np.nan
-        print('Warning - output is Nan')
+    # Get output from a run and delete output files
+    warpx_out = read_sim_output(task.workdir)
+
+    # Excluding results - NAN - from runs where beam was lost
+    if (warpx_out[0] != warpx_out[0]):
+        print(task.workdir, ' output led to NAN values')
```
Comment (ax3l) [inline Tools/LibEnsemble/warpx_simf.py:93]: Just an idea to forward the reason for NaNs to the output.
```suggestion
    # Excluding results - NaN - from runs where beam was lost
    if (warpx_out[0] != warpx_out[0]):
        print(task.workdir, ' output led to NaN values (beam loss)')
```

**Review by ax3l — APPROVED** [review]
LGTM! :+1:

File: Tools/LibEnsemble/warpx_simf.py (hunk 2 above)
Comment (ax3l) [reply Tools/LibEnsemble/warpx_simf.py:93]: (totally optional, can be a follow-up if it's useful or be ignored)

Comment (LDAmorim) [reply Tools/LibEnsemble/warpx_simf.py:93]: Sorry @ax3l , I only saw this now. Yes, it's a good idea.
I'll include it in a next PR!

Comment (ax3l) [reply Tools/LibEnsemble/warpx_simf.py:93]: Sounds good - the looked so minor to me that I though you can choose and I'll merge this useful update already :)

Comment (LDAmorim) [reply Tools/LibEnsemble/warpx_simf.py:93]: There are prints next to the ifs that test either "fails" beam loss or no plotfiles.
I am preparing a branch for this then.
What do you think of:
`print(task.workdir, ' output led to NaN values (beam was lost or run did not finish)')`

Comment (ax3l) [reply Tools/LibEnsemble/warpx_simf.py:93]: looks great, feel free to break the string into multiple lines as well.

**Pull request merged**